prov/gni/src/gnix_buddy_allocator.c

```diff
@@ -0,0 +1,353 @@
+		│ BASE │ len  │ min │ max │ nlists │ LISTS │
+		└──────┴──────┴─────┴─────┴────────┴───────┘
+ * The LISTS pointer points to an array of dlist structures, each containing a
+ * head pointer to the begging of a free list.  Note that the first element of
+ * LISTS is a pointer to the head of the "min block size" free list, the second
+ * element is the head of the "min * 2 block size" free list and so on.
+
+ 		Node format as stored in a free block:
+		┌──────┬──────┬──────────────────────┐
+		│ NEXT │ PREV │ Remaining free bytes │
+		└──────┴──────┴──────────────────────┘
+ * Each NEXT and PREV pointer is stored in the first 16 bytes of the free block.
+ * This means that there is a hard limit of 16 bytes on the minimum block size.
+ *
+ *		Bitmap layout with a min block size of 16:
+ 		┌──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┬──┐
+		│16│16│16│16│..│32│32│32│32│..│64│64│64│64│..│
+		└──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┴──┘
+ * Each number above (16, 32, and 64) represents a block size that can
+ * be split. The first two "32" blocks are redundantly mapped by the the first
+ * 64 block, all blocks are redundant mappings of the next block size.
+ *
+ * When a block is split or allocated, its bit will be set in the bitmap.
+ * When a block is coalesced or free'd its bit will be reset in the bitmap.
+ * This improves the performance of coalescing blocks by being able to tell
+ * whether a given block's buddy block is allocated or split in O(c).
+ *
+ * TODO: dlist_insert_sorted for fragmentation reduction.
+ * TODO: Lock in __gnix_buddy_split and allow __gnix_buddy_find_block to run
+ * concurrently.
+ */
+
+#include "gnix_buddy_allocator.h"
+
+static inline int __gnix_buddy_create_lists(handle_t *alloc_handle)
+{
+	size_t i, offset = 0;
+
+	alloc_handle->nlists = (size_t) log2(alloc_handle->max /
+					     (double) MIN_BLOCK_SIZE) + 1;
+	alloc_handle->lists = calloc(1, sizeof(struct dlist_entry) *
+				     alloc_handle->nlists);
+
+	if (!alloc_handle->lists) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Could not create buddy allocator lists.\n");
+		return -FI_ENOMEM;
+	}
+
+	for (i = 0; i < alloc_handle->nlists; i++) {
+		dlist_init(alloc_handle->lists + i);
+	}
+
+	/* Insert free blocks of size max in sorted order into last list */
+	for (i = 0; i < alloc_handle->len / alloc_handle->max; i++) {
+		dlist_insert_tail(alloc_handle->base + offset,
+				  alloc_handle->lists +
+				  alloc_handle->nlists - 1);
+		offset += alloc_handle->max;
+	}
+
+	return FI_SUCCESS;
+}
+
+/**
+ * Split a block in list "j" until list "i" is reached.
+ */
+static inline void __gnix_buddy_split(handle_t *alloc_handle, size_t j, size_t i)
+{
+	void *tmp = NULL;
+
+	dlist_remove(tmp = alloc_handle->lists[j].next);
+
+	_gnix_set_bit(&alloc_handle->bitmap,
+		      BITMAP_INDEX(tmp, alloc_handle->base, MIN_BLOCK_SIZE,
+				   OFFSET(MIN_BLOCK_SIZE, j)));
+
+
+	/* Split the block until we reach list "i" */
+	for (j--; j > i; j--) {
```

Comment:
I'm not sure I understand.  Perhaps I need to rethink the bitmap layout.  From what I understand the buddy allocator should be managing a base block of at most 32MB.

Assuming the largest block size is 16MB and the smallest block size is 16 bytes, the largest difference between j and i would be 20 since:
i = 0 (the index of the 16byte block list)
j = 20 (the index of the 16MB block list).

> On Jan 7, 2016, at 5:45 PM, James Swaro <notifications@github.com> wrote:
> 
> In prov/gni/src/gnix_buddy_allocator.c:
> 
> > +/**
> > + * Split a block in list "j" until list "i" is reached.
> > + */
> > +static inline void __gnix_buddy_split(handle_t *alloc_handle, size_t j, size_t i)
> > +{
> > +	void *tmp = NULL;
> > +
> > +	dlist_remove(tmp = alloc_handle->lists[j].next);
> > +
> > +	_gnix_set_bit(&alloc_handle->bitmap,
> > +		      BITMAP_INDEX(tmp, alloc_handle->base, MIN_BLOCK_SIZE,
> > +				   OFFSET(MIN_BLOCK_SIZE, j)));
> > +
> > +
> > +	/* Split the block until we reach list "i" */
> > +	for (j--; j > i; j--) {
> Could you speculate what the most common difference between j and i will be? 1, 2, 4? If the value is high enough, you might find a linear set of the bitset to be prohibitively slow.
> 
> —
> Reply to this email directly or view it on GitHub.
>